ludwig/schema/config_object.py

```diff
@@ -0,0 +1,501 @@
+    """OutputFeatures is a container for all output features."""
+
+    pass
+
+
+@dataclass(repr=False)
+class ModelConfig(BaseMarshmallowConfig):
+    """This class is the implementation of the config object that replaces the need for a config dictionary
+    throughout the project."""
+
+    def __init__(self, config_dict: dict):
+
+        # ===== Backwards Compatibility =====
+        upgraded_config_dict = self._upgrade_config(config_dict)
+
+        # ===== Initialize Top Level Config Sections =====
+
+        # Since 'ecd' is the default model type, we set it initially here.
+        # If a user specifies a model type however, it will be overridden later.
+        self.model_type: str = MODEL_ECD
+        self.input_features: InputFeaturesContainer = InputFeaturesContainer()
+        self.output_features: OutputFeaturesContainer = OutputFeaturesContainer()
+        self.combiner: BaseCombinerConfig = ConcatCombinerConfig()
+        self.trainer: BaseTrainerConfig = ECDTrainerConfig()
+        self.preprocessing: PreprocessingConfig = PreprocessingConfig()
+        self.defaults: DefaultsConfig = copy.deepcopy(DefaultsConfig())
+
+        # ===== Set User Defined Global Defaults =====
+        if DEFAULTS in upgraded_config_dict:
+            self._set_attributes(self.defaults, upgraded_config_dict[DEFAULTS])
+
+        # ===== Features =====
+        self._set_feature_column(upgraded_config_dict)
+        self._set_proc_column(upgraded_config_dict)
+        self._initialize_input_features(upgraded_config_dict[INPUT_FEATURES])
+        self._set_input_features(upgraded_config_dict[INPUT_FEATURES])
+        self._initialize_output_features(upgraded_config_dict[OUTPUT_FEATURES])
+        self._set_output_features(upgraded_config_dict[OUTPUT_FEATURES])
+
+        # ===== Model Type =====
+        if MODEL_TYPE in upgraded_config_dict:
+            if upgraded_config_dict[MODEL_TYPE] == MODEL_GBM:
+                self.model_type = MODEL_GBM
+                self.trainer = GBMTrainerConfig()
+                if (
+                    TYPE in upgraded_config_dict.get(TRAINER, {})
+                    and upgraded_config_dict[TRAINER][TYPE] != "lightgbm_trainer"
+                ):
+                    raise ValidationError("GBM Model trainer must be of type: 'lightgbm_trainer'")
+
+                for feature in self.input_features.to_dict().keys():
+                    feature_cls = getattr(self.input_features, feature)
+                    if feature_cls.type == BINARY:
+                        feature_cls.encoder = BinaryPassthroughEncoderConfig()
+                    elif feature_cls.type in [CATEGORY, NUMBER]:
+                        feature_cls.encoder = PassthroughEncoderConfig()
+                    else:
+                        raise ValidationError(
+                            "GBM Models currently only support Binary, Category, and Number " "features"
+                        )
```

Comment:
nit: Move this to a separate function, i.e. `set_gbm_attributes()`

Reply:
Good call! Done!